wen-restart/src/heaviest_fork_aggregate.rs

```diff
@@ -0,0 +1,560 @@
+            observed_stake: record.total_active_stake,
+            shred_version: record.shred_version as u16,
+        };
+        Ok(self.aggregate(converted_record))
+    }
+
+    pub(crate) fn aggregate(
+        &mut self,
+        received_heaviest_fork: RestartHeaviestFork,
+    ) -> Option<HeaviestForkRecord> {
+        let total_stake = self.epoch_stakes.total_stake();
+        let from = &received_heaviest_fork.from;
+        let sender_stake = Self::validator_stake(&self.epoch_stakes, from);
+        if sender_stake == 0 {
+            warn!(
+                "Gossip should not accept zero-stake RestartLastVotedFork from {:?}",
+                from
+            );
+            return None;
+        }
+        if received_heaviest_fork.shred_version != self.my_shred_version {
+            warn!(
+                "Gossip should not accept RestartLastVotedFork with different shred version {} from {:?}",
+                received_heaviest_fork.shred_version, from
+            );
+            return None;
+        }
+        let record = HeaviestForkRecord {
+            slot: received_heaviest_fork.last_slot,
+            bankhash: received_heaviest_fork.last_slot_hash.to_string(),
+            total_active_stake: received_heaviest_fork.observed_stake,
+            shred_version: received_heaviest_fork.shred_version as u32,
+            wallclock: received_heaviest_fork.wallclock,
+        };
+        if let Some(old_heaviest_fork) = self
+            .heaviest_forks
+            .insert(*from, received_heaviest_fork.clone())
+        {
+            if old_heaviest_fork == received_heaviest_fork {
+                return None;
+            } else {
+                let entry = self
+                    .block_stake_map
+                    .get_mut(&(
+                        old_heaviest_fork.last_slot,
+                        old_heaviest_fork.last_slot_hash,
+                    ))
+                    .unwrap();
+                info!(
+                    "{:?} Replacing old heaviest fork from {:?} with {:?}",
+                    from, old_heaviest_fork, received_heaviest_fork
+                );
+                *entry = entry.saturating_sub(sender_stake);
+            }
+        }
+        let entry = self
+            .block_stake_map
+            .entry((
+                received_heaviest_fork.last_slot,
+                received_heaviest_fork.last_slot_hash,
+            ))
+            .or_insert(0);
+        *entry = entry.saturating_add(sender_stake);
+        self.active_peers.insert(*from);
+        if received_heaviest_fork.observed_stake as f64 / total_stake as f64
+            >= self.supermajority_threshold
+        {
+            self.active_peers_seen_supermajority.insert(*from);
+        }
+        if !self
+            .active_peers_seen_supermajority
+            .contains(&self.my_pubkey)
+            && self.total_active_stake() as f64 / total_stake as f64 >= self.supermajority_threshold
```

Comment:
Should this `supermajority_threshold` also account for the malicious 5%

Reply:
Yes, since this threshold is passed in and only used to calculate percent of stake seeing majority, I'll change it on the caller side.